File: 1721.swapping-nodes-in-a-linked-list.py
#
# @lc app=leetcode id=1721 lang=python3
#
# [1721] Swapping Nodes in a Linked List
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional 
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = left = right = first = ListNode(next = head)
        for i in range(k):
            right = right.next
            #now we have reached the first node 
            first = first.next
        
        while right.next:
            right = right.next
            #now left is the node previous to the kth node from the end
            left = left.next
        
        #exchange the value of the two nodes
        val = first.val
        second = left.next 
        first.val = second.val
        second.val = val
        return dummy.next

# @lc code=end

File: 1721.swapping-nodes-in-a-linked-list.java
/*
 * @lc app=leetcode id=1721 lang=java
 *
 * [1721] Swapping Nodes in a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapNodes(ListNode head, int k) {
        // if(head.next == null) return head;
        // int cnt = 0;
        // ListNode traverse = head;
        // while(traverse != null){
        //     cnt++;
        //     traverse = traverse.next;
        // }
        // ListNode fir = head, sec = head;
        // for(int i = 0; i < k -1; i++){
        //     fir = fir.next;
        // }
        // for(int j = 0; j < cnt - k; j++){
        //     sec = sec.next;
        // }
        // int val = fir.val;
        // fir.val = sec.val;
        // sec.val = val;
        // return head;


        //TODO complete it in one pass
        //! method of using fast and slow pointers
        // save redundant list traverse and complete in one pass 

        if(head == null || head.next == null) return head;

        ListNode first = head; 
        ListNode second = head;

        k -= 1;

        ListNode cur = head;
        while(cur.next != null){
            cur = cur.next;
            k--;
            if(k == 0){
                first = cur;
            }
            if(k < 0){
                second = second.next;
            }
        }
        int val = first.val;
        first.val  = second.val;
        second.val = val;
        return head;
    }
}
// @lc code=end

